R/eval.R

```diff
@@ -169,22 +174,17 @@ evaluate_top_level_expression <- function(exprs,
   }
   eHandler <- function(cnd) {
     handle_output()
-    if (use_try) {
-      cnd <- reset_call(cnd)
-      watcher$errored()
-      watcher$push(cnd)
-      output_handler$error(cnd)
-    }
-  }
-
-  if (use_try) {
-    handle <- function(code) {
-      tryCatch(code, error = function(err) {
-        list(value = NULL, visible = FALSE)
-      })
+    
+    cnd <- reset_call(cnd)
+    watcher$push(cnd)
+    
+    if (on_error == "continue") {
+      invokeRestart("eval_continue")
+    } else if (on_error == "stop") {
+      invokeRestart("eval_stop")
+    } else {
+      invokeRestart("eval_error", cnd)
```

Comment:
It took me a little while to understand why we need this — but without it the error will propagate up to other error handlers that might be registered on the stack. In particular, this means that `expect_snapshot(evaluate("stop('1')", stop_on_error = 2))` includes unexpected text because the error handler set up by the evaluate call in `expect_snapshot()` ends up capturing the propagated value.